File: src/utils/logger.py
"""
Logging configuration for the application.
"""
import logging
import os
import sys
from pathlib import Path
from typing import Optional

def setup_logger(log_level: Optional[str] = None, 
                log_file: Optional[str] = None) -> None:
    """
    Configure application logging.
    
    Args:
        log_level: Logging level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        log_file: Path to log file
    """
    log_level = log_level or os.environ.get("LOG_LEVEL", "INFO")
    numeric_level = getattr(logging, log_level.upper(), logging.INFO)
    
    # Create logs directory if not exists and log_file is not provided
    if not log_file:
        log_dir = Path(__file__).parent.parent.parent / "logs"
        log_dir.mkdir(exist_ok=True)
        log_file = str(log_dir / "app.log")
    
    # Configure root logger
    logging.basicConfig(
        level=numeric_level,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
        handlers=[
            logging.FileHandler(log_file),
            logging.StreamHandler(sys.stdout)
        ]
    )
    
    # Set logging level for external libraries
    logging.getLogger("openai").setLevel(logging.WARNING)
    logging.getLogger("httpx").setLevel(logging.WARNING)
    logging.getLogger("googleapiclient").setLevel(logging.WARNING)
    
    logger = logging.getLogger(__name__)
    logger.info(f"Logging configured with level {log_level}")
    logger.info(f"Log file: {log_file}")

File: src/utils/logger.js
/**
 * Logger utility for the application
 */

class Logger {
  constructor() {
    this.logLevel = process.env.LOG_LEVEL || 'info';
    this.levels = {
      error: 0,
      warn: 1,
      info: 2,
      debug: 3
    };
    
    // Add file logging option
    this.logToFile = process.env.LOG_TO_FILE === 'true';
    this.logFilePath = process.env.LOG_FILE_PATH || './logs/app.log';
    
    // Initialize file logging if enabled
    if (this.logToFile) {
      this._initFileLogging();
    }
  }

  /**
   * Check if the given log level should be logged
   */
  shouldLog(level) {
    return this.levels[level] <= this.levels[this.logLevel];
  }

  /**
   * Format the log message
   */
  formatMessage(level, message, meta = {}) {
    const timestamp = new Date().toISOString();
    const metaString = Object.keys(meta).length ? JSON.stringify(meta) : '';
    return `[${timestamp}] [${level.toUpperCase()}] ${message} ${metaString}`.trim();
  }

  /**
   * Initialize file logging
   * @private
   */
  _initFileLogging() {
    const fs = require('fs');
    const path = require('path');
    
    // Create logs directory if it doesn't exist
    const logDir = path.dirname(this.logFilePath);
    if (!fs.existsSync(logDir)) {
      fs.mkdirSync(logDir, { recursive: true });
    }
  }
  
  /**
   * Write log to file
   * @private
   */
  _writeToFile(formattedMessage) {
    if (!this.logToFile) return;
    
    const fs = require('fs');
    fs.appendFileSync(this.logFilePath, formattedMessage + '\n');
  }
  
  /**
   * Clear log files
   */
  clearLogs() {
    if (this.logToFile) {
      const fs = require('fs');
      fs.writeFileSync(this.logFilePath, '');
      this.info('Log file cleared');
    }
  }
  
  /**
   * Set log level dynamically
   */
  setLogLevel(level) {
    if (this.levels[level] !== undefined) {
      this.logLevel = level;
      this.info(`Log level set to: ${level}`);
      return true;
    }
    return false;
  }

  /**
   * Log an error message
   */
  error(message, meta = {}) {
    if (this.shouldLog('error')) {
      const formattedMessage = this.formatMessage('error', message, meta);
      console.error(formattedMessage);
      this._writeToFile(formattedMessage);
    }
  }

  /**
   * Log a warning message
   */
  warn(message, meta = {}) {
    if (this.shouldLog('warn')) {
      const formattedMessage = this.formatMessage('warn', message, meta);
      console.warn(formattedMessage);
      this._writeToFile(formattedMessage);
    }
  }

  /**
   * Log an info message
   */
  info(message, meta = {}) {
    if (this.shouldLog('info')) {
      const formattedMessage = this.formatMessage('info', message, meta);
      console.info(formattedMessage);
      this._writeToFile(formattedMessage);
    }
  }

  /**
   * Log a debug message
   */
  debug(message, meta = {}) {
    if (this.shouldLog('debug')) {
      const formattedMessage = this.formatMessage('debug', message, meta);
      console.debug(formattedMessage);
      this._writeToFile(formattedMessage);
    }
  }
}

module.exports = new Logger();
